File: common/src/main/java/com/aliyun/emr/rss/common/network/protocol/Message.java
/*
 * Licensed to the Apache Software Foundation (ASF) under one or more
 * contributor license agreements.  See the NOTICE file distributed with
 * this work for additional information regarding copyright ownership.
 * The ASF licenses this file to You under the Apache License, Version 2.0
 * (the "License"); you may not use this file except in compliance with
 * the License.  You may obtain a copy of the License at
 *
 *    http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package com.aliyun.emr.rss.common.network.protocol;

import java.nio.ByteBuffer;

import com.google.common.base.Objects;
import io.netty.buffer.ByteBuf;
import io.netty.buffer.Unpooled;

import com.aliyun.emr.rss.common.network.buffer.ManagedBuffer;
import com.aliyun.emr.rss.common.network.buffer.NettyManagedBuffer;

/** An on-the-wire transmittable message. */
public abstract class Message implements Encodable {
  private ManagedBuffer body;

  protected Message() {
    this(null);
  }

  protected Message(ManagedBuffer body) {
    this.body = body;
  }

  /** Used to identify this request type. */
  public abstract Type type();

  /** An optional body for the message. */
  public ManagedBuffer body() {
    return body;
  }

  public void setBody(ByteBuf buf) {
    this.body = new NettyManagedBuffer(buf);
  }

  /** Whether the body should be copied out in frame decoder. */
  public boolean needCopyOut() {
    return false;
  }

  protected boolean equals(Message other) {
    return Objects.equal(body, other.body);
  }

  public ByteBuffer toByteBuffer() {
    // Allow room for encoded message, plus the type byte
    ByteBuf buf = Unpooled.buffer(encodedLength() + 1);
    buf.writeByte(type().id());
    encode(buf);
    assert buf.writableBytes() == 0 : "Writable bytes remain: " + buf.writableBytes();
    return buf.nioBuffer();
  }

  /** Preceding every serialized Message is its type, which allows us to deserialize it. */
  public enum Type implements Encodable {
    UnkownType(-1),
    ChunkFetchRequest(0),
    ChunkFetchSuccess(1),
    ChunkFetchFailure(2),
    RpcRequest(3),
    RpcResponse(4),
    RpcFailure(5),
    OpenStream(6),
    StreamHandle(7),
    OneWayMessage(9),
    PushData(11),
    PushMergedData(12);

    private final byte id;

    Type(int id) {
      assert id < 128 : "Cannot have more than 128 message types";
      this.id = (byte) id;
    }

    public byte id() {
      return id;
    }

    @Override
    public int encodedLength() {
      return 1;
    }

    @Override
    public void encode(ByteBuf buf) {
      buf.writeByte(id);
    }

    public static Type decode(ByteBuf buf) {
      byte id = buf.readByte();
      switch (id) {
        case 0:
          return ChunkFetchRequest;
        case 1:
          return ChunkFetchSuccess;
        case 2:
          return ChunkFetchFailure;
        case 3:
          return RpcRequest;
        case 4:
          return RpcResponse;
        case 5:
          return RpcFailure;
        case 6:
          return OpenStream;
        case 7:
          return StreamHandle;
        case 9:
          return OneWayMessage;
        case 11:
          return PushData;
        case 12:
          return PushMergedData;
        case -1:
          throw new IllegalArgumentException("User type messages cannot be decoded.");
        default:
          throw new IllegalArgumentException("Unknown message type: " + id);
      }
    }
  }

  public static Message decode(Type msgType, ByteBuf in) {
    return decode(msgType, in, true);
  }

  public static Message decode(Type msgType, ByteBuf in, boolean decodeBody) {
    switch (msgType) {
      case ChunkFetchRequest:
        return ChunkFetchRequest.decode(in);

      case ChunkFetchSuccess:
        return ChunkFetchSuccess.decode(in, decodeBody);

      case ChunkFetchFailure:
        return ChunkFetchFailure.decode(in);

      case RpcRequest:
        return RpcRequest.decode(in, decodeBody);

      case RpcResponse:
        return RpcResponse.decode(in, decodeBody);

      case RpcFailure:
        return RpcFailure.decode(in);

      case OpenStream:
        return OpenStream.decode(in);

      case StreamHandle:
        return StreamHandle.decode(in);

      case OneWayMessage:
        return OneWayMessage.decode(in, decodeBody);

      case PushData:
        return PushData.decode(in, decodeBody);

      case PushMergedData:
        return PushMergedData.decode(in, decodeBody);

      default:
        throw new IllegalArgumentException("Unexpected message type: " + msgType);
    }
  }

  public static Message decode(ByteBuffer buffer) {
    ByteBuf buf = Unpooled.wrappedBuffer(buffer);
    Type type = Type.decode(buf);
    return decode(type, buf);
  }
}

File: cpp/celeborn/network/Message.cpp
/*
 * Licensed to the Apache Software Foundation (ASF) under one or more
 * contributor license agreements.  See the NOTICE file distributed with
 * this work for additional information regarding copyright ownership.
 * The ASF licenses this file to You under the Apache License, Version 2.0
 * (the "License"); you may not use this file except in compliance with
 * the License.  You may obtain a copy of the License at
 *
 *    http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

#include "celeborn/network/Message.h"
#include "celeborn/protocol/Encoders.h"

namespace celeborn {
namespace network {
Message::Type Message::decodeType(uint8_t typeId) {
  switch (typeId) {
    case 0:
      return CHUNK_FETCH_REQUEST;
    case 1:
      return CHUNK_FETCH_SUCCESS;
    case 2:
      return CHUNK_FETCH_FAILURE;
    case 3:
      return RPC_REQUEST;
    case 4:
      return RPC_RESPONSE;
    case 5:
      return RPC_FAILURE;
    case 6:
      return OPEN_STREAM;
    case 7:
      return STREAM_HANDLE;
    case 9:
      return ONE_WAY_MESSAGE;
    case 11:
      return PUSH_DATA;
    case 12:
      return PUSH_MERGED_DATA;
    case 13:
      return REGION_START;
    case 14:
      return REGION_FINISH;
    case 15:
      return PUSH_DATA_HAND_SHAKE;
    case 16:
      return READ_ADD_CREDIT;
    case 17:
      return READ_DATA;
    case 18:
      return OPEN_STREAM_WITH_CREDIT;
    case 19:
      return BACKLOG_ANNOUNCEMENT;
    case 20:
      return TRANSPORTABLE_ERROR;
    case 21:
      return BUFFER_STREAM_END;
    case 22:
      return HEARTBEAT;
    default:
      CELEBORN_FAIL("Unknown message type " + std::to_string(typeId));
  }
}

std::atomic<long> Message::currRequestId_ = 0;

std::unique_ptr<memory::ReadOnlyByteBuffer> Message::encode() const {
  int bodyLength = body_->remainingSize();
  int encodedLength = internalEncodedLength();
  int headerLength =
      sizeof(int32_t) + sizeof(uint8_t) + sizeof(int32_t) + encodedLength;
  auto buffer = memory::ByteBuffer::createWriteOnly(headerLength);
  buffer->write<int32_t>(encodedLength);
  buffer->write<uint8_t>(type_);
  buffer->write<int32_t>(bodyLength);
  internalEncodeTo(*buffer);
  auto header = memory::ByteBuffer::toReadOnly(std::move(buffer));
  auto combinedFrame = memory::ByteBuffer::concat(*header, *body_);
  return std::move(combinedFrame);
}

std::unique_ptr<Message> Message::decodeFrom(
    std::unique_ptr<memory::ReadOnlyByteBuffer>&& data) {
  int32_t encodedLength = data->read<int32_t>();
  uint8_t typeId = data->read<uint8_t>();
  int32_t bodyLength = data->read<int32_t>();
  CELEBORN_CHECK_EQ(encodedLength + bodyLength, data->remainingSize());
  Type type = decodeType(typeId);
  switch (type) {
    case RPC_RESPONSE:
      return RpcResponse::decodeFrom(std::move(data));
    case RPC_FAILURE:
      return RpcFailure::decodeFrom(std::move(data));
    case CHUNK_FETCH_SUCCESS:
      return ChunkFetchSuccess::decodeFrom(std::move(data));
    case CHUNK_FETCH_FAILURE:
      return ChunkFetchFailure::decodeFrom(std::move(data));
    default:
      CELEBORN_FAIL("unsupported Message decode type " + std::to_string(type));
  }
}

int RpcRequest::internalEncodedLength() const {
  return sizeof(long) + sizeof(int32_t);
}

void RpcRequest::internalEncodeTo(memory::WriteOnlyByteBuffer& buffer) const {
  buffer.write<long>(requestId_);
  buffer.write<int32_t>(body_->remainingSize());
}

std::unique_ptr<RpcResponse> RpcResponse::decodeFrom(
    std::unique_ptr<memory::ReadOnlyByteBuffer>&& data) {
  long requestId = data->read<long>();
  data->skip(4);
  auto result = std::make_unique<RpcResponse>(requestId, std::move(data));
  return result;
}

std::unique_ptr<RpcFailure> RpcFailure::decodeFrom(
    std::unique_ptr<memory::ReadOnlyByteBuffer>&& data) {
  long requestId = data->read<long>();
  int strLen = data->read<int>();
  CELEBORN_CHECK_EQ(data->remainingSize(), strLen);
  std::string errorString = data->readToString(strLen);
  return std::make_unique<RpcFailure>(requestId, std::move(errorString));
}

std::unique_ptr<ChunkFetchSuccess> ChunkFetchSuccess::decodeFrom(
    std::unique_ptr<memory::ReadOnlyByteBuffer>&& data) {
  protocol::StreamChunkSlice streamChunkSlice =
      protocol::StreamChunkSlice::decodeFrom(*data);
  return std::make_unique<ChunkFetchSuccess>(streamChunkSlice, std::move(data));
}

std::unique_ptr<ChunkFetchFailure> ChunkFetchFailure::decodeFrom(
    std::unique_ptr<memory::ReadOnlyByteBuffer>&& data) {
  protocol::StreamChunkSlice streamChunkSlice =
      protocol::StreamChunkSlice::decodeFrom(*data);
  int strLen = data->read<int>();
  CELEBORN_CHECK_EQ(data->remainingSize(), strLen);
  std::string errorString = data->readToString(strLen);
  return std::make_unique<ChunkFetchFailure>(
      streamChunkSlice, std::move(errorString));
}

PushData::PushData(
    long requestId,
    uint8_t mode,
    const std::string& shuffleKey,
    const std::string& partitionUniqueId,
    std::unique_ptr<memory::ReadOnlyByteBuffer> body)
    : Message(PUSH_DATA, std::move(body)),
      requestId_(requestId),
      mode_(mode),
      shuffleKey_(shuffleKey),
      partitionUniqueId_(partitionUniqueId) {}

PushData::PushData(const PushData& other)
    : Message(PUSH_DATA, other.body()),
      requestId_(other.requestId_),
      mode_(other.mode_),
      shuffleKey_(other.shuffleKey_),
      partitionUniqueId_(other.partitionUniqueId_) {}

int PushData::internalEncodedLength() const {
  return sizeof(long) + sizeof(uint8_t) + protocol::encodedLength(shuffleKey_) +
      protocol::encodedLength(partitionUniqueId_);
}

void PushData::internalEncodeTo(memory::WriteOnlyByteBuffer& buffer) const {
  buffer.write<long>(requestId_);
  buffer.write<uint8_t>(mode_);
  protocol::encode(buffer, shuffleKey_);
  protocol::encode(buffer, partitionUniqueId_);
}
} // namespace network
} // namespace celeborn
